cmd/neofs-lens/internal/storage/sanity.go

```diff
@@ -0,0 +1,224 @@
+				sh.p = peapod.New(subCfg.Path(), subCfg.Perm(), peapodCfg.FlushInterval())
+
+				var compressCfg compression.Config
+				err := compressCfg.Init()
+				common.ExitOnErr(cmd, common.Errf("failed to init compression config: %w", err))
+
+				sh.p.SetCompressor(&compressCfg)
+			case fstree.Type:
+				fstreeCfg := fstreeconfig.From((*config.Config)(subCfg))
+				sh.fsT = fstree.New(
+					fstree.WithPath(subCfg.Path()),
+					fstree.WithPerm(subCfg.Perm()),
+					fstree.WithDepth(fstreeCfg.Depth()),
+					fstree.WithNoSync(fstreeCfg.NoSync()),
+				)
+			}
+		}
+
+		common.ExitOnErr(cmd, common.Errf("open metabase: %w", sh.m.Open(true)))
+		common.ExitOnErr(cmd, common.Errf("open peapod: %w", sh.p.Open(true)))
+		common.ExitOnErr(cmd, common.Errf("open fstree: %w", sh.fsT.Open(true)))
+
+		// metabase.Open(true) does not set it mode to RO somehow
+		common.ExitOnErr(cmd, common.Errf("moving metabase in readonly mode", sh.m.SetMode(mode.ReadOnly)))
+
+		common.ExitOnErr(cmd, common.Errf("init metabase: %w", sh.m.Init()))
+		common.ExitOnErr(cmd, common.Errf("init peapod: %w", sh.p.Init()))
+		common.ExitOnErr(cmd, common.Errf("init fstree: %w", sh.fsT.Init()))
+
+		shards = append(shards, sh)
+
+		return nil
+	})
+	common.ExitOnErr(cmd, common.Errf("reading config: %w", err))
+
+	defer func() {
+		for _, sh := range shards {
+			_ = sh.m.Close()
+			_ = sh.p.Close()
+			_ = sh.fsT.Close()
+		}
+	}()
+
+	for _, sh := range shards {
+		idRaw, err := sh.m.ReadShardID()
+		if err != nil {
+			cmd.Printf("reading shard id: %s; skip this shard\n", err)
+			continue
+		}
+
+		id := base58.Encode(idRaw)
+		cmd.Printf("Checking %s shard\n", id)
+
+		err = checkShard(cmd, sh)
+		if err != nil {
+			if errors.Is(err, context.Canceled) {
+				return
+			}
+
+			cmd.Printf("%s shard check: %s\n", id, err)
+			continue
+		}
+	}
+}
+
+func checkShard(cmd *cobra.Command, sh storageShard) error {
+	var mPrm meta.ListPrm
+	mPrm.SetCount(1024)
+
+	for {
+		listRes, err := sh.m.ListWithCursor(mPrm)
+		if err != nil {
+			if errors.Is(err, meta.ErrEndOfListing) {
+				return nil
+			}
+
+			return fmt.Errorf("listing objects in meta: %w", err)
+		}
+
+		for _, obj := range listRes.AddressList() {
+			select {
+			case <-cmd.Context().Done():
+				return cmd.Context().Err()
+			default:
+			}
+
+			var sIDPrm meta.StorageIDPrm
+			sIDPrm.SetAddress(obj.Address)
+
+			sIDRes, err := sh.m.StorageID(sIDPrm)
+			if err != nil {
+				return fmt.Errorf("reading storage ID in meta: %w", err)
+			}
+
+			var mGet meta.GetPrm
+			mGet.SetAddress(obj.Address)
+
+			getRes, err := sh.m.Get(mGet)
+			if err != nil {
+				return fmt.Errorf("reading object in meta: %w", err)
+			}
+
+			header := *getRes.Header()
+
+			if sIDRes.StorageID() == nil {
+				err = checkObject(header, sh.fsT)
+			} else {
+				err = checkObject(header, sh.p)
+			}
+			if err != nil {
+				cmd.Printf("%s object check: %s\n", obj.Address, err)
+				continue
```

Comment:
in theory, we may encounter static error (e.g. peapod died due to whatever reason) which we'll be repeated for all next objects. In this caseб if we have ~1KK objects - stdout is gonna be a mess. Have you considered dividing errors into normal and fatal?

not really critical

Reply:
> Have you considered dividing errors into normal and fatal?

no, do you have a practical suggestion?

Reply:
storages may return typed logical [errors](https://pkg.go.dev/github.com/nspcc-dev/neofs-node@v0.42.1/pkg/local_object_storage/util/logicerr), mb continue only if this happens and break otherwise?